File: python/streaming/window_network_wordcount.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, print_function, division, unicode_literals

import sys

from pyspark import SparkConf, SparkContext
from pyspark.storagelevel import StorageLevel
from pyspark.streaming.context import StreamingContext


if __name__ == '__main__':
    conf = SparkConf().setAppName('Windows Network Word Count')
    sc = SparkContext(conf=conf)
    ssc = StreamingContext(sc, 5)
    ssc.checkpoint("checkpoint")

    lines = ssc.socketTextStream(sys.argv[1], int(sys.argv[2]), StorageLevel.MEMORY_AND_DISK_SER)
    word_counts = (
        lines.flatMap(lambda l: l.split(' '))
        .map(lambda w: (w, 1L))
        .reduceByKeyAndWindow(
            lambda x, y: x + y,
            lambda x, y: x - y,
            15
        )
    )

    word_counts.pprint()

    ssc.start()
    ssc.awaitTermination()

File: scala/src/main/scala/streaming/WindowNetworkWordCount.scala
package streaming

import org.apache.spark._
import org.apache.spark.storage.StorageLevel
import org.apache.spark.streaming.StreamingContext._
import org.apache.spark.streaming._

object WindowNetworkWordCount {

  def main(args: Array[String]) {
    run(args(0), args(1).toInt)
  }

  def run(hostname: String, port: Int): Unit = {
    val conf = new SparkConf().setAppName("Window Network Word Count")
    conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")

    val ssc = new StreamingContext(conf, Seconds(5))
    ssc.checkpoint("checkpoint")

    val lines = ssc.socketTextStream(hostname, port, StorageLevel.MEMORY_AND_DISK_SER)
    val wordCounts = lines.flatMap(_.split(" "))
      .map(_.toLowerCase)
      .map((_, 1L))
      .reduceByKeyAndWindow(_ + _, _ - _, Seconds(15))

    wordCounts.print()

    ssc.start()
    ssc.awaitTermination()
  }
}
